File: core/src/com/multiplayer/Courrier.java
package com.multiplayer;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Net;
import com.badlogic.gdx.net.Socket;
import com.badlogic.gdx.net.SocketHints;
import com.tiles.Player;
import com.utils.Multiplayer;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Courrier {
    private final Socket sendingSocket;

    public boolean getAnswer() {
        return answer;
    }

    public Socket getReceivingSocket() {
        return receivingSocket;
    }

    private Socket receivingSocket;

    boolean answer = false; // This will be used to get answers from the server
    private final String id;

    // sendingSocket: La socket pour envoyer des messages au serveur
    // receivingSocket: La socket pour recevoir des messages du serveur

    private final ClientListener clientListener;

    public Courrier(String id, int port, String ip) throws ServerNotReachedException {
        this.id = id;
        SocketHints socketHints = new SocketHints();
        socketHints.connectTimeout = 5000;
        socketHints.sendBufferSize = 1024;
        socketHints.receiveBufferSize = 1024;
        try {
            sendingSocket = Gdx.net.newClientSocket(Net.Protocol.TCP, ip, port, socketHints);
            sendMessage("connected " + ip);
        } catch (Exception e) {
            throw new ServerNotReachedException("Server not found");
        }
        String waitForIt = null;
        try {
            waitForIt = (new BufferedReader(new InputStreamReader(sendingSocket.getInputStream()))).readLine();
            if (waitForIt.equals("server rejected you"))
                throw new ServerNotReachedException("Username is already taken");
            receivingSocket = Gdx.net.newClientSocket(Net.Protocol.TCP, ip, port, socketHints);
            waitForIt = (new BufferedReader(new InputStreamReader(receivingSocket.getInputStream()))).readLine();
        } catch (IOException e) {
            e.printStackTrace();
        }

        sendObject(Multiplayer.me);
        clientListener = new ClientListener(Multiplayer.key, receivingSocket);
        clientListener.startThread();

    }

    public void sendMessage(String message) {
        try {
            //println("Client: Sending $message")
            sendingSocket.getOutputStream().write((id + " " + message + "\n").getBytes());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void sendObject(BigButton toSend) {
        sendMessage("sending BigButton");
        //ObjectOutputStream(sendingSocket.outputStream).write(toSend.serialize().toByteArray())
        try {
            sendingSocket.getOutputStream().write(((toSend.serialize() + "\n").getBytes()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void sendObject(Player toSend) {
        sendMessage("sending Player");
        //ObjectOutputStream(sendingSocket.outputStream).write(toSend.serialize().toByteArray())
        try {
            sendingSocket.getOutputStream().write(((Multiplayer.mapper.writeValueAsString(toSend) + " \n").getBytes()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void killThread() {
        clientListener.killThread();
    }
}

File: core/src/com/multiplayer/Courrier.kt
package com.multiplayer

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Net
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.net.Socket
import com.badlogic.gdx.net.SocketHints
import com.tiles.Player
import com.utils.Multiplayer
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

// Le but de cette classe est d'éviter de recréer une socket à chaque fois
// Et de faciliter l'envois de messages.

class Courrier(val id: String, port: Int, ip: String) {
    val sendingSocket: Socket
    val receivingSocket: Socket
    val socketHints: SocketHints
    var answer = false // This will be used to get answers from the server

    // sendingSocket: La socket pour envoyer des messages au serveur
    // receivingSocket: La socket pour recevoir des messages du serveur

    init {
        socketHints = SocketHints()
        socketHints.connectTimeout = 0
        socketHints.sendBufferSize = 1024
        socketHints.receiveBufferSize = 1024
        sendingSocket = Gdx.net.newClientSocket(Net.Protocol.TCP, ip, port, socketHints)
        sendMessage("connected $ip")
        var waitForIt = BufferedReader(InputStreamReader(sendingSocket.inputStream)).readLine()
        receivingSocket = Gdx.net.newClientSocket(Net.Protocol.TCP, ip, port, socketHints)
        waitForIt = BufferedReader(InputStreamReader(receivingSocket.inputStream)).readLine()
        sendObject(Multiplayer.me)
        ClientListener(Multiplayer.key, receivingSocket).startThread()
    }

    fun sendMessage(message: String) {
        try {
            //println("Client: Sending $message")
            sendingSocket.getOutputStream().write(("$id $message \n").toByteArray())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun sendObject(toSend: BigButton) {
        sendMessage("sending BigButton")
        //ObjectOutputStream(sendingSocket.outputStream).write(toSend.serialize().toByteArray())
        sendingSocket.getOutputStream().write(((toSend.serialize() + "\n").toByteArray()))
    }

    fun sendObject(toSend: Player) {
        sendMessage("sending Player")
        //ObjectOutputStream(sendingSocket.outputStream).write(toSend.serialize().toByteArray())
        sendingSocket.getOutputStream().write(((Multiplayer.mapper.writeValueAsString(toSend) + " \n").toByteArray()))
    }

}